### PR TITLE
Picture puzzle activity

### PR DESCRIPTION
**Pull request by Raxerz**

### CODE REVIEW

File: src/app/introduction/introduction.component.scss [hunk 1]
```diff
@@ -0,0 +1,162 @@
+@import "../styles/components/_variables";
```
Comment (yukie1412) [inline src/app/introduction/introduction.component.scss]: this is very different from the `introduction.component.scss` in https://github.com/systers/PC-Prep-Kit/pull/25/files

File: src/app/introduction/picture-puzzle.component.ts [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+        this.webcamButtonText = 'Take Photo';
+        this.checkProgress();
+    }
+
+    upload() {
+        const formData: any = new FormData();
+        const files: Array<File> = this.filesToUpload;
+        if (this.webcamState === this.webcamStates.CAPTURED) {
+            const camData = {base64: this._canvas.toDataURL('image/jpeg')}
+            this._http.post('http://localhost:3000/api/uploadCam', camData)
```
Comment (yukie1412) [inline src/app/introduction/picture-puzzle.component.ts]: localhost?

File: src/app/introduction/picture-puzzle.component.ts [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+        if (this.webcamStream) {
+            if (this.webcamStream.stop) {
+                this.webcamStream.stop();
+            } else {
+                this.webcamStream.getTracks().forEach(function(track) {track.stop()})
+            }
+        }
+        this.webcamState = this.webcamStates.PAGE_LOAD;
+        this.webcamButtonText = 'Take Photo';
+        const file = document.getElementById('file').click();
```
Comment (yukie1412) [inline src/app/introduction/picture-puzzle.component.ts]: where is it used?

File: src/app/introduction/picture-puzzle.component.ts [hunk 4]
```diff
@@ -0,0 +1,384 @@
...
+
+    ngOnInit() {
+        this.webcamButtonText = 'Take Photo';
+        this.checkProgress();
+    }
+
+    upload() {
+        const formData: any = new FormData();
+        const files: Array<File> = this.filesToUpload;
+        if (this.webcamState === this.webcamStates.CAPTURED) {
```
Comment (yukie1412) [inline src/app/introduction/picture-puzzle.component.ts]: a huge chunk of `if-else` is repeated here, which can be simplified & reduced

File: src/app/introduction/picture-puzzle.component.ts [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+        const file = document.getElementById('file').click();
+    }
+
+    handleInputChange(e) {
+        const file = e.dataTransfer ? e.dataTransfer.files[0] : e.target.files[0];
+        const pattern = /image-*/;
+        const reader = new FileReader();
+
+        if (!file.type.match(pattern)) {
+            alert('invalid format');
```
Comment (yukie1412) [inline src/app/introduction/picture-puzzle.component.ts]: `alert`, `confirm` are extremely ugly (blocks js execution & stops everything entirely) & is not suggested by js.
you should use modals intead

File: src/app/introduction/picture-puzzle.component.ts [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+
+    snapshot() {
+        this.webcamState = this.webcamStates.CAPTURED;
+        if (this.webcamStream.stop) {
+            this.webcamStream.stop();
+        } else {
+            this.webcamStream.getTracks().forEach(function(track) { track.stop() })
+        }
+        this.webcamButtonText = 'Take another photo';
+        this._pieceWidth = Math.floor(this._width / this.PUZZLE_DIFFICULTY)
```
Comment (yukie1412) [inline src/app/introduction/picture-puzzle.component.ts]: how is it different from `onImage()`?

File: src/app/introduction/picture-puzzle.component.ts [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+        const reader = e.target;
+        this.imageSrc = reader.result;
+        this.loaded = true;
+        this._img = new Image();
+        this._img.src = this.imageSrc;
+        this.imageLoaded = true;
+        this._renderer.listen(this._img, 'load', (event) => this.onImage(event));
+    }
+
+    onImage(e) {
```
Comment (yukie1412) [inline src/app/introduction/picture-puzzle.component.ts]: where are you using `e`?

File: src/app/introduction/picture-puzzle.component.ts [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+        this._newImg.src = this._canvas.toDataURL('image/png');
+        this._renderer.listen(this._newImg, 'load', (event) => this.buildPieces(event));
+    }
+
+    buildPieces(e) {
+        let i;
+        let piece;
+        let xPos = 0;
+        let yPos = 0;
+        for (i = 0; i < this.PUZZLE_DIFFICULTY * this.PUZZLE_DIFFICULTY; i++) {
```
Comment (yukie1412) [inline src/app/introduction/picture-puzzle.component.ts]: spacing here conflicts with https://github.com/systers/PC-Prep-Kit/pull/25/files
pick one

File: src/app/introduction/picture-puzzle.component.ts [hunk 9]
```diff
@@ -0,0 +1,384 @@
...
+        this.puzzleState = 'Shuffle';
+        this._pieces = this.shuffleArray(this._pieces);
+        this._stage.clearRect(0, 0, this._puzzleWidth, this._puzzleHeight);
+        for (i = 0;  i < this._pieces.length; i++) {
+            piece = this._pieces[i];
+            piece.xPos = xPos;
+            piece.yPos = yPos;
+            this._stage.drawImage(this._newImg, piece.sx, piece.sy, this._pieceWidth, this._pieceHeight, xPos, yPos, this._pieceWidth, this._pieceHeight);
+            this._stage.strokeRect(xPos, yPos, this._pieceWidth, this._pieceHeight);
+            xPos += this._pieceWidth;
```
Comment (yukie1412) [inline src/app/introduction/picture-puzzle.component.ts]: duplicates with `buildPieces`

honestly, if you are reusing the same/similar chunk of codes multiple times, you should consider utility functions

File: src/app/introduction/picture-puzzle.component.ts [hunk 10]
```diff
@@ -0,0 +1,384 @@
...
+    checkProgress() {
+        this._dashboardService.getProgressStatus().subscribe(response => {
+            this._activity = response.activity;
+            this._stage = response.stage;
+            if (this._stage >= 1 && this._activity >= 3) {
+                this.activityComplete = true;
+            }
+        });
+    }
+}
```
Comment (yukie1412) [inline src/app/introduction/picture-puzzle.component.ts]: try to clean this file up slightly, especially reduce the amount of repeated code

good reads: http://www.itiseezee.com/?p=119
this is slightly extreme, but does have few good points & practices

File: src/app/introduction/picture-puzzle.component.ts (hunk 8 above)
Comment (Raxerz) [inline src/app/introduction/picture-puzzle.component.ts]: tslint (Typescript lint) throws errors if I change the spacing for this.

File: src/app/introduction/picture-puzzle.component.ts (hunk 9 above)
Comment (Raxerz) [inline src/app/introduction/picture-puzzle.component.ts]: They are two different functions and both are essential in developing this activity.
The local variables are named the same in both these functions so I'm guessing that is probably why they look alike.

File: routes/api.js [hunk 11]
```diff
@@ -108,4 +118,44 @@ router.get('/mailpcpolicy', authenticationHelpers.isAuthOrRedirect, (req, res) =
...
 
+/**
+ * [destination description]
+ * @param  {Object} req       Request object
+ * @param  {Object} file      File object
+ * @param  {Function} cb      Callback function
+ */
+const storage = multer.diskStorage({
+    destination: function(req, file, cb) {
+        cb(null, './uploads/')
```
Comment (yukie1412) [inline routes/api.js]: semi-colon

File: src/app/app.module.ts [hunk 12]
```diff
@@ -28,6 +28,11 @@ import { APIService } from './services/api.service';
 import { RegService } from './services/reg.service';
 import { NavbarService } from './services/navbar.service';
 
+import { SharedDataService } from './services/shared.data.service';
+import { IntroductionComponent } from './introduction/introduction.component'
+import { HighlightActivityComponent } from './introduction/highlight-activity.component'
```
Comment (yukie1412) [inline src/app/app.module.ts]: semi-colon (consider adding it to the linting files)

File: src/app/introduction/introduction.component.scss (hunk 1 above)
Comment (yukie1412) [inline src/app/introduction/introduction.component.scss]: so, which of these 2 `introduction.component.scss` is correct?
you are only going to cause merge conflicts at this rate

File: src/app/introduction/picture-puzzle.component.html [hunk 13]
```diff
@@ -0,0 +1,29 @@
+<div class="row">
+    <div class="col-lg-8 col-lg-offset-3">
+        <h2>Take a picture of yourself, someone or something important to you</h2>
+    </div>
+</div>
+<div class="row">
+    <div class="col-md-3 col-md-offset-3">
+        <video #video width="480" height="360" autoplay [hidden]='webcamState!=webcamStates.OPENED'></video>
```
Comment (yukie1412) [inline src/app/introduction/picture-puzzle.component.html]: `!==`

File: src/app/introduction/picture-puzzle.component.html [hunk 14]
```diff
@@ -0,0 +1,29 @@
...
+        <canvas width="480" height="360" [hidden]='webcamState===webcamStates.OPENED' id="canvas"></canvas>
+    </div>
+    <div class="col-md-3 col-md-offset-3">
+        <input type="file" name="file" id="file" accept="image/*" (change)="handleInputChange($event)">
+        <button class="btn btn-block btn-danger btn-lg" *ngIf='webcamState===webcamStates.PAGE_LOAD' (click)="openWebcam()">{{webcamButtonText}}</button>
```
Comment (yukie1412) [inline src/app/introduction/picture-puzzle.component.html]: as far as i can tell, there are only 3 states to `webcamStates`, names `PAGE_LOAD`, `CAPTURED`, & `OPENED`.
then, why not `*ngIf='webcamState===webcamStates.OPENED'` & `*ngIf='webcamState!==webcamStates.OPENED'` instead?

File: src/app/introduction/picture-puzzle.component.ts (hunk 9 above)
Comment (yukie1412) [inline src/app/introduction/picture-puzzle.component.ts]: yes, but a huge chunk of the code are exactly the same. have you ever considered utility/auxiliary/helper functions to reduce the amount of code duplication?

File: src/app/introduction/picture-puzzle.component.html (hunk 14 above)
Comment (Raxerz) [inline src/app/introduction/picture-puzzle.component.html]: Can you please elaborate a little more on this @yukie1412

File: src/app/introduction/picture-puzzle.component.ts (hunk 9 above)
Comment (Raxerz) [inline src/app/introduction/picture-puzzle.component.ts]: I've eliminated some of the superficial code duplicates. The other pieces of code that contain similar lines of code are in a different sequence and the sequence matters.
Also I'm not sure if using utilility functions for eliminating local variable duplicates across functions will serve a good purpose.

File: src/app/introduction/introduction.component.scss (hunk 1 above)
Comment (yukie1412) [inline src/app/introduction/introduction.component.scss]: see corresponding comments in the other PR

File: src/app/introduction/picture-puzzle.component.html [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+    </div>
+</div>
+<div class="row">
+    <div class="col-md-3 col-md-offset-3">
+        <video #video width="480" height="360" autoplay [hidden]='webcamState !== webcamStates.OPENED'></video>
+        <canvas width="480" height="360" [hidden]='webcamState === webcamStates.OPENED' id="canvas"></canvas>
+    </div>
+    <div class="col-md-3 col-md-offset-3">
+        <input type="file" name="file" id="file" accept="image/*" (change)="handleInputChange($event)">
+        <button class="btn btn-block btn-danger btn-lg" *ngIf='webcamState === webcamStates.PAGE_LOAD' (click)="openWebcam()">{{webcamButtonText}}</button>
```
Comment (yukie1412) [inline src/app/introduction/picture-puzzle.component.html]: line 13 & 14 can be simplified to:
```
<button class="btn btn-block btn-danger btn-lg" *ngIf='webcamState !== webcamStates.OPENED' (click)="openWebcam()">{{webcamButtonText}}</button>
```

File: src/app/introduction/picture-puzzle.component.ts (hunk 9 above)
Comment (yukie1412) [inline src/app/introduction/picture-puzzle.component.ts]: i would actually have a function to replace the following lines:

```
getPosition (posn) {  // could be more descriptive - what position?
    posn.x += this._pieceWidth;
    if (posn.x >= this._puzzleWidth) {
        posn.x = 0;
        posn.y += this._pieceHeight;
    }
    return posn;
}
```

can could be reused in `buildPieces()` as well

File: src/app/introduction/picture-puzzle.component.ts [hunk 16]
```diff
@@ -0,0 +1,370 @@
...
+                    this._stage.globalAlpha = .4;
+                    this._stage.fillStyle = this.PUZZLE_HOVER_TINT;
+                    this._stage.fillRect(this._currentDropPiece.xPos, this._currentDropPiece.yPos, this._pieceWidth, this._pieceHeight);
+                    this._stage.restore();
+                }
+            }
+        }
+        this._stage.save();
+        this._stage.globalAlpha = .6;
+        this._stage.drawImage(this._newImg, this._currentPiece.sx, this._currentPiece.sy, this._pieceWidth, this._pieceHeight, this._mouse.x - (this._pieceWidth / 2), this._mouse.y - (this._pieceHeight / 2), this._pieceWidth, this._pieceHeight);
```
Comment (yukie1412) [inline src/app/introduction/picture-puzzle.component.ts]: i actually going to variables for ` this._mouse.x - (this._pieceWidth / 2)` & `this._mouse.y - (this._pieceHeight / 2)` to reduce the amount of calculation needed

File: src/app/introduction/picture-puzzle.component.ts [hunk 17]
```diff
@@ -0,0 +1,370 @@
...
+        this._stage.restore();
+        this._stage.strokeRect( this._mouse.x - (this._pieceWidth / 2), this._mouse.y - (this._pieceHeight / 2), this._pieceWidth, this._pieceHeight);
+    }
+
+    pieceDropped(e) {
+        this._mousemoveListener();
+        this._mouseupListener();
+        if (this._currentDropPiece != null) {
```
Comment (yukie1412) [inline src/app/introduction/picture-puzzle.component.ts]: `!==`

File: src/app/introduction/picture-puzzle.component.ts [hunk 18]
```diff
@@ -0,0 +1,370 @@
...
+        this._dashboardService.getProgressStatus().subscribe(response => {
+            this._activity = response.activity;
+            this._stage = response.stage;
+            if (this._stage >= 1 && this._activity >= 3) {
+                this.activityComplete = true;
+            }
+        });
+    }
+
+    customAlert(title, msg, type) {
```
Comment (yukie1412) [inline src/app/introduction/picture-puzzle.component.ts]: should this be global?

File: src/app/services/dashboard.service.ts [hunk 19]
```diff
@@ -8,27 +8,64 @@ import 'rxjs/Rx';
 @Injectable()
 export class DashboardService {
 
+    private _localStorageKey = environment.localStorageKey;
+
     private _getUserInfo = environment.baseURL + environment.apiEndpoint + 'getUserInfo';
```
Comment (yukie1412) [inline src/app/services/dashboard.service.ts]: why not a const for `environment.baseURL + environment.apiEndpoint`?

File: src/app/introduction/introduction.component.scss [hunk 20]
```diff
@@ -0,0 +1,147 @@
...
+
+.rectangle-speech-border {
+    position: relative;
+    padding: 50px;
+    width: 100%;
+    margin: 1em 0 3em;
+    border: 2px solid black;
+    color: #333;
+    background: $white;
+    -webkit-border-radius: 50px;
```
Comment (yukie1412) [inline src/app/introduction/introduction.component.scss]: `@include border-radius(50px)`?

File: src/app/introduction/picture-puzzle.component.html [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+    </div>
+</div>
+<div class="row">
+    <div class="col-md-3 col-md-offset-3">
+        <video #video width="480" height="360" autoplay [hidden]='webcamState !== webcamStates.OPENED'></video>
+        <canvas width="480" height="360" [hidden]='webcamState === webcamStates.OPENED' id="canvas"></canvas>
+    </div>
+    <div class="col-md-3 col-md-offset-3">
+        <input type="file" name="file" id="file" class="hidden" accept="image/*" (change)="handleInputChange($event)">
+        <button class="btn btn-block btn-danger btn-lg" *ngIf='webcamState!==webcamStates.OPENED' (click)="openWebcam()">{{webcamButtonText}}</button>
```
Comment (yukie1412) [inline src/app/introduction/picture-puzzle.component.html]: spacing

File: src/app/introduction/picture-puzzle.component.ts [hunk 22]
```diff
@@ -0,0 +1,351 @@
...
+                    this._stage.globalAlpha = .4;
+                    this._stage.fillStyle = this.PUZZLE_HOVER_TINT;
+                    this._stage.fillRect(this._currentDropPiece.xPos, this._currentDropPiece.yPos, this._pieceWidth, this._pieceHeight);
+                    this._stage.restore();
+                }
+            }
+        }
+        this._stage.save();
+        this._stage.globalAlpha = .6;
+        const mouseXPos = this._mouse.x - (this._pieceWidth / 2);
```
Comment (yukie1412) [inline src/app/introduction/picture-puzzle.component.ts]: 305-308 are the same as 256-259, worth another refactor

**Comment by yukie1412** [comment]
resolve the conflicts

**Comment by Raxerz** [comment]
I have resolved the conflicts @yukie1412. Can you please check?

**Pull request merged**